FF1Lib/ShuffleAstos.cs

```diff
@@ -0,0 +1,123 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Diagnostics;
+using RomUtilities;
+
+namespace FF1Lib
+{
+
+	public partial class FF1Rom : NesRom
+	{
+		public void ShuffleAstos(Flags flags, MT19337 rng)
+		{
+			const int NpcTalkOffset = 0x390D3;
+			const int NpcTalkSize = 2;
+
+			// NPC pool to swap Asto with
+			// We can't do NPC that checks for flags and give items, because there's not enough space in Talk_Astos
+			// So Lefein, ElfPrince and vanilla Sarda are out
+			// However ElfDoc is fine and Early Sarda too
+			List<ObjectId> npcpool = new List<ObjectId> { ObjectId.Astos, ObjectId.Bahamut, ObjectId.CanoeSage, ObjectId.CubeBot, ObjectId.ElfDoc,
+			ObjectId.Fairy, ObjectId.King, ObjectId.Matoya, ObjectId.Nerrick, ObjectId.Princess2, ObjectId.Smith,
+			ObjectId.Titan, ObjectId.Unne };
+
+			var scriptStandardNPCItemTrade = Blob.FromHex("5693");
+			var scriptTalk_Astos = Blob.FromHex("3893");
+
+			// Add Sarda if the early flag is on
+			if (flags.EarlySarda ?? false) npcpool.Add(ObjectId.Sarda);
+
+			// Select random npc
+			ObjectId newastos = npcpool.PickRandom(rng);
+
+			// If Astos, we're done here
+			if (newastos == ObjectId.Astos) return;
+
+			// If not get NPC talk routine, get NPC object
+			var talkscript = Get(NpcTalkOffset + (byte)newastos * NpcTalkSize, 2);
+			var talkvalue = Get(MapObjOffset + (byte)newastos * MapObjSize, 4);
+
+			// Switch astos to TalkScripts.StandardNPCItemTrade, set scriptvalue to crown Item.Crown;
+			Put(NpcTalkOffset + (byte)ObjectId.Astos * NpcTalkSize, scriptStandardNPCItemTrade);
+			Put(MapObjOffset + (byte)ObjectId.Astos * MapObjSize, Blob.FromHex("02")); // 0x02 = Crown
+
+			// Swtich NPC to Astos
+			Put(NpcTalkOffset + (byte)newastos * NpcTalkSize, scriptTalk_Astos);
+
+			// Universal fix to Talk_Astos: ObjectID is this object, not Astos
+			Put(0x39338 + 14, Blob.FromHex("A416"));
+
+			// Change dialog a bit for non-item giving NPCs
+			Put(0x285EF + 18, Blob.FromHex("00"));
+
+			if (newastos == ObjectId.Titan)
+			{
+				// Check required item, don't try to give any item
+				Put(0x39338, Blob.FromHex("AD2960F016EAEAEAEAEAEAEAEAEA"));
+
+				// Change dialog
+				Put(0x39338 + 24, Blob.FromHex("A912"));
+			}
+			else if (newastos == ObjectId.Bahamut)
+			{   // Butcher Talk_Astos to make it work with Bahamut, and also modify DoClassChange

```

Comment:
So I think other features of the rando definitely mess with things like DoClassChange. I'm not sure if this would bork in conjunction with those.

Reply:
yeah, iirc it was messed with for Nones, check out the file FF1Lib/asm/0E_95AE_DoClassChange.asm

Reply:
It seems to be the only instance and the randomizer always call PartyComposition(), so it should be safe to assume it's in place. I'll need to move ShuffleAstos() after PartyComposition() in the randomizer so it doesn't overwrite the bahamut fix.